### ISSUE TITLE
.mjs file from library incorrectly picked up during SSR

### ISSUE THREAD

**Issue by akheron** [gatsbyjs/gatsby#22599]
## Description

I'm trying to use [react-sticky-nav](https://github.com/iiroj/react-sticky-nav/) in a Gatsby site. After importing the library, the Gatsby build fails with this error:

```
 ERROR #98123  WEBPACK

Generating SSR bundle failed

Can't import the named export 'forwardRef' from non EcmaScript module (only default export is available)

File: node_modules/react-sticky-nav/index.mjs
```

### Steps to reproduce

```
git clone https://github.com/akheron/gatsby-ssr-mjs-test
cd gatsby-ssr-mjs-test
npm install
npm run build
```

### Expected result

It should compile fine.

### Actual result

I get the error mentioned in the Description section of this issue.

### Environment

    System:
      OS: Linux 5.5 Arch Linux
      CPU: (4) x64 Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
      Shell: 5.0.16 - /bin/bash
    Binaries:
      Node: 13.11.0 - /usr/bin/node
      Yarn: 1.22.4 - /usr/bin/yarn
      npm: 6.14.3 - /usr/bin/npm
    Languages:
      Python: 3.8.2 - /usr/bin/python
    Browsers:
      Firefox: 74.0
    npmPackages:
      gatsby: ^2.19.45 => 2.20.7 

### My own analysis

I started from the gatsby hello world starter, installed react-sticky-nav, and changed `src/pages/index.js` to use it:

```
import React from "react"
import Navigation from "react-sticky-nav"

export default () => (
  <div>
    <Navigation />
    <div>Hello world!</div>
  </div>
)
```

react-sticky-nav's package.json is built as both CommonJS moudle and ES module. Its package.json has:
```
  "main": "index.js",
  "module": "index.mjs",
```

index.mjs has:
```
import React,{memo,forwardRef,useState,useRef,useCallback,useEffect}from'react';
```

React is a CommonJS-only module, so this kind of importing is not valid in NodeJS. An ES module can only import the default export from a CommonJS module. 

However, I think it works with webpack, i.e. webpack's import/require tracking supports importing non-default exports from CommonJS modules, as well as mixing import and require.

The question is, why is `node_modules/react-sticky-nav/index.mjs` picked up by Gatsby when doing SSR instead of `node_modules/react-sticky-nav/index.js`.

### Workaround

Forcing the use of `node_modules/react-sticky-nav/index.js` fixes this issue.

gatsby-node.js:
```
exports.onCreateWebpackConfig = ({ actions }) => {
  actions.setWebpackConfig({
    resolve: {
      alias: {
        'react-sticky-nav': path.resolve(
          __dirname,
          'node_modules/react-sticky-nav/index.js'
        ),
      },
    },
  })
}
```

**Comment by Js-Brecht** [comment]
Resolving `.mjs` first is not a Gatsby thing; it's a webpack thing:

> https://webpack.js.org/configuration/resolve/
> 
> Meanwhile, a Node.js application bundled by webpack will first try to resolve using the file in the module field.

During SSR, the [webpack target is set to `node`](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/utils/webpack.config.js#L467), which will cause webpack to resolve using the `module` field, if it exists.  If I remove that field from the `package.json`, and then remove the extension from `main`, then webpack will resolve in the order of [`resolve.extensions`](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/bootstrap/index.js#L458).

I'm not convinced this is a problem with Gatsby at all.  IMO, the impetus should be on the author(s) of `react-sticky-nav` to make sure their package is using compatible standards... React is cjs, so the way they are calling it is not compatible until React begins to align their shipped code with ES standards.  

Perhaps other people with more clout will have a different opinion, though.  I'm sure there's a way Gatsby could work around this, but is it worth it?  🤷‍♂

**Comment by vladar** [comment]
@Js-Brecht Excellent explanation of what is going on there. Thanks!

@akheron Also check out this suggestion (and maybe the whole thread): https://github.com/webpack/webpack/issues/7482#issuecomment-394955916

tl;dr this should be addressed in the `react-sticky-nav` package.

**Comment by LekoArts** [comment]
Thank you for opening this!

We're marking this issue as answered and closing it for now but please feel free to comment here if you would like to continue this discussion. We also recommend heading over to our [communities](https://www.gatsbyjs.org/contributing/community/#where-to-get-support) if you have questions that are not bug reports or feature requests. We hope we managed to help and thank you for using Gatsby!
